### DISCUSSION TITLE
DCIM/Devices: Modules not added

### DISCUSSION THREAD

**Discussion by heidl89** [netbox-community/netbox#17861] [Help Wanted!] [answered]
### Deployment Type

Self-hosted

### Triage priority

I'm a NetBox Labs customer

### NetBox Version

v4.1.4

### Python Version

3.12

### Steps to Reproduce

1. Create a custom script "new device"
2. Add some data to "Device" Object
3. device.save()
4. add interface, assign ip to interface
5. settattr -> primary ip
6. device.save()

Within the second save, modules (power outlets, console ports) were not added in the gui. 
The Change Log of the device says the modules are generated....

Important: 
If i skip the second device.save() the modules were created...
But the ip assignment (settattr) is not added...

My code: 

def run (self, data, commit):

    if not data['gps_lat']:   
            data['gps_lat'] = 0
    if not data['gps_long']:
            data['gps_long'] = 0

    
    
    device = Device(
            name = data['device_name'],
            device_type = data['device_type'],
            role = data['device_role'],
            status = data['status'],
            latitude = data['gps_lat'],
            longitude = data['gps_long'],
            site_id = data['site'].id
     )
     
    if data['location']:             
            device.location = data['location']
    
    device.full_clean()
    device.save()

    
    interface = Interface(
            device = device,
            name = "vmgmt0",
            type = InterfaceTypeChoices.TYPE_VIRTUAL,
            enabled = True,
            mgmt_only = True
    )

    interface.full_clean()
    interface.save()
    
    ip = IPAddress(
        address = data['ipaddress'],
        status = IPAddressStatusChoices.STATUS_ACTIVE,
        assigned_object = interface
        
    )
    ip.full_clean()
    ip.save()
    
    setattr(device,"primary_ip4", ip)

    device.save()
         
    self.log_success(f"Created new Device {device}")

### Expected Behavior

Make Topic Console Ports, Power Ports visible in Device View

### Observed Behavior

These Objects were generated, but arent visible...

![device_view](https://github.com/user-attachments/assets/97ae0c53-66f6-48e8-b9cc-75995386f29a)
![device_changelog](https://github.com/user-attachments/assets/a52354cf-7831-4806-9445-f52d8812d93a)

**Comment by candlerb** [comment]
Aside: duplicate of #17842.

> If i skip the second device.save() the modules were created...

That's useful information, because the modules are created by the *first* save (when the device itself is created). Therefore, what's likely happening is that the second save is resetting the cached module counts to zero.

Try reloading the device with [refresh_from_db()](https://docs.djangoproject.com/en/5.0/ref/models/instances/#refreshing-objects-from-database) before setting the primary_ip4 and saving it back.

```
device.refresh_from_db()
device.primary_ip4 = ip
device.full_clean()
device.save()
```

Or use update() to set only the primary_ip4:

```
Device.objects.filter(pk=device.pk).update(primary_ip4=ip)
```

Aside: why use `setattr`? Why not `device.primary_ip4 = ip` ?

**Reply by heidl89** [reply]
@candlerb: thx for your advice: 

I've closed my other post. 

The reason for using settattr was because of this issue.... I thougt this would fix my problem.. :)

**Comment by candlerb** [comment, marked as answer]
I can reproduce your problem in nbshell:

```
>>> d = Device(device_type=DeviceType.objects.get(model="RB4011iGS+RM"), role=DeviceRole.objects.get(name="Server"), site=Site.objects.get(name="Test"))
>>> d.full_clean()
>>> d.save()
>>> d.description="hello world"
>>> d.full_clean()
>>> d.save()
>>> d.pk
107
```

I go to `/dcim/devices/107/` in the UI, I see the 'interfaces' tab but not the `power ports` or `console ports` tab.

```
>>> [(a, getattr(d,a)) for a in dir(d) if "_count" in a]
[('console_port_count', 0), ('console_server_port_count', 0), ('device_bay_count', 0), ('front_port_count', 0), ('interface_count', 0), ('interfaces_count', 11), ('inventory_item_count', 0), ('module_bay_count', 0), ('power_outlet_count', 0), ('power_port_count', 0), ('rear_port_count', 0)]
```

Notice it's a bit weird to have both "interfaces_count" (which is correct) and "interface_count" (which is wrong). I think `interfaces_count` has virtual-chassis-specific logic and doesn't use the cached count mechanism:

```
    @property
    def interfaces_count(self):
        return self.vc_interfaces().count()
```

Then if I run in a different window:

```
# /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py calculate_cached_counts
```

Back in the first window:

```
>>> d.refresh_from_db()
>>> [(a, getattr(d,a)) for a in dir(d) if "_count" in a]
[('console_port_count', 1), ('console_server_port_count', 0), ('device_bay_count', 0), ('front_port_count', 0), ('interface_count', 11), ('interfaces_count', 11), ('inventory_item_count', 0), ('module_bay_count', 0), ('power_outlet_count', 0), ('power_port_count', 1), ('rear_port_count', 0)]
```

Now the counts are correct. And the GUI displays the missing tabs.

The general solution (to avoid running calculate_cached_counts) is to refresh the device from the db after creating it:

```
>>> d = Device(device_type=DeviceType.objects.get(model="RB4011iGS+RM"), role=DeviceRole.objects.get(name="Server"), site=Site.objects.get(name="Test"))
>>> d.full_clean()
>>> d.save()
>>> d.refresh_from_db()             # ADD THIS
>>> d.description="hello world"
>>> d.full_clean()
>>> d.save()
>>> d.pk
108
```

This new device has the correct tabs visible.

**Reply by candlerb** [reply]
I think I found why "Interfaces" are handled differently in the GUI to other component types. In `netbox/dcim/views.py`:

```
@register_model_view(Device, 'powerports', path='power-ports')
class DevicePowerPortsView(DeviceComponentsView):
    child_model = PowerPort
    table = tables.DevicePowerPortTable
    filterset = filtersets.PowerPortFilterSet
    filterset_form = forms.PowerPortFilterForm
    template_name = 'dcim/device/powerports.html'
    tab = ViewTab(
        label=_('Power Ports'),
        badge=lambda obj: obj.power_port_count,              <<< CACHED COUNT
        permission='dcim.view_powerport',
        weight=570,
        hide_if_empty=True
    )
```

But:

```
@register_model_view(Device, 'interfaces')
class DeviceInterfacesView(DeviceComponentsView):
    child_model = Interface
    table = tables.DeviceInterfaceTable
    filterset = filtersets.InterfaceFilterSet
    filterset_form = forms.InterfaceFilterForm
    template_name = 'dcim/device/interfaces.html'
    tab = ViewTab(
        label=_('Interfaces'),
        badge=lambda obj: obj.vc_interfaces().count(),       <<< CALCULATED ON DEMAND
        permission='dcim.view_interface',
        weight=520,
        hide_if_empty=True
    )
```

And this in turn is because interfaces are dealt with differently on a virtual chassis master, whereas other components are not.

**Reply by heidl89** [reply]
@candlerb

Thx for your help and time spending to analyse this :)
Your solution works fine...
Thx for this, you re my hero ;)
